clientv3/watch.go

```diff
@@ -41,3 +46,480 @@ type WatchResponse struct {
+	select {
+	case w.stopc <- struct{}{}:
+	case <-w.donec:
+	}
+	<-w.donec
+	return <-w.errc
+}
+
+// watch posts a watch request to run() and waits for a new watcher channel
+func (w *watcher) watch(ctx context.Context, key, prefix string, rev int64) <-chan WatchResponse {
+	retc := make(chan chan WatchResponse, 1)
+	wr := &watchRequest{ctx, key, prefix, rev, retc}
+	// submit request
+	select {
+	case w.reqc <- wr:
+	case <-wr.ctx.Done():
+		return nil
+	case <-w.donec:
+		return nil
+	}
+	// recieve channel
+	select {
+	case ret := <-retc:
+		return ret
+	case <-ctx.Done():
+		return nil
+	case <-w.donec:
+		return nil
+	}
+}
+
+func (w *watcher) addStream(resp *pb.WatchResponse, pendingReq *watchRequest) {
+	if resp.WatchId == -1 {
+		// failed; no channel
+		pendingReq.retc <- nil
+		return
+	}
+
+	ret := make(chan WatchResponse)
+	ws := &watcherStream{
+		initReq: *pendingReq,
+		id:      resp.WatchId,
+		outc:    ret,
+		recvc:   make(chan *WatchResponse),
+	}
+
+	w.mu.Lock()
+	w.streams[ws.id] = ws
+	w.mu.Unlock()
+
+	// send messages to subscriber
+	go w.serveStream(ws)
+	// listen for ctx cancels
+	w.ctxc <- struct{}{}
+
+	// pass back the subscriber channel for the watcher
+	pendingReq.retc <- ret
+}
+
+// closeStream closes the watcher resources and removes it
+func (w *watcher) closeStream(ws *watcherStream) {
+	// cancels request stream; subscriber receives nil channel
+	close(ws.initReq.retc)
+	// close subscriber's channel
+	close(ws.outc)
+	// shutdown serveStream
+	close(ws.recvc)
+	delete(w.streams, ws.id)
+}
+
+// run is the root of the goroutines for managing a watcher client
+func (w *watcher) run() {
+	defer func() {
+		w.mu.Lock()
+		for _, ws := range w.streams {
+			w.closeStream(ws)
+		}
+		w.mu.Unlock()
+		close(w.donec)
+		w.cancel()
+	}()
+
+	// start a stream with the etcd grpc server
+	wc, wcerr := w.newWatchClient()
+	if wcerr != nil {
+		w.errc <- wcerr
+		return
+	}
+
+	// close stream when its ctx is done
+	go w.serveCtxs()
+
+	var pendingReq, failedReq *watchRequest
+	curReqC := w.reqc
+	cancelSet := make(map[int64]struct{})
+
+	for {
+		select {
+		// Watch() requested
+		case pendingReq = <-curReqC:
+			// no more watch requests until there's a response
+			curReqC = nil
+			if err := wc.Send(pendingReq.toPB()); err == nil {
+				// pendingReq now waits on w.respc
+				break
+			}
+			failedReq = pendingReq
+		// New events from the watch client
+		case pbresp := <-w.respc:
+			switch {
+			case pbresp.Canceled:
+				delete(cancelSet, pbresp.WatchId)
+			case pbresp.Compacted:
+				w.mu.Lock()
+				if ws, ok := w.streams[pbresp.WatchId]; ok {
+					w.closeStream(ws)
+				}
+				w.mu.Unlock()
+			case pbresp.Created:
+				if pendingReq != nil {
+					// response to pending req, try to add
+					w.addStream(pbresp, pendingReq)
+					pendingReq = nil
+					curReqC = w.reqc
+				}
+			default:
+				// dispatch to appropriate watch stream
+				if ok := w.dispatchEvent(pbresp); ok {
```

Comment:
after a careful look, i feel this is probably OK since we are controlling this internally.